xbmc/cores/VideoPlayer/VideoPlayerAudio.cpp

```diff
@@ -531,9 +543,27 @@ bool CVideoPlayerAudio::ProcessDecoderOutput(DVDAudioFrame &audioframe)
   }
 
 
+  if (m_synctype == SYNC_DISCON)
   {
     double syncerror = m_audioSink.GetSyncError();
-    if (m_synctype == SYNC_DISCON && fabs(syncerror) > DVD_MSEC_TO_TIME(m_disconAdjustTimeMs))
+
+    if (m_disconLearning)
+    {
+      const double syncErr = fabs(syncerror);
+      if (syncErr > DVD_MSEC_TO_TIME(m_disconAdjustTimeMs))
+        m_disconAdjustTimeMs = DVD_TIME_TO_MSEC(syncErr);
```

Comment:
The sanity check is below. Maybe we should do something on top. Imagine the following: On a setup where disks are in standby, where readreate is currently limited - it might very well happen on startup that we run into a 150 ms discontinuity or a 200 ms one. As we don't correct this (we are still learning) this non-corrected offset will stay in the code forever. Only if additionally a 100 ms issue happens later on, we correct.

I would change the code and immediately correct this peak.

```
m_disconAdjustTimeMs = DVD_TIME_TO_MSEC(syncErr);
if (m_disconAdjustTimeMs > 100)
{
   m_disconAdjustTimeMs = 100;
   m_disconLearning = false;
   m_disconTimer.SetExpired();
   CLog::LogF(LOGINFO, "Algorithm cancelled as of large discon error of {} ms", DVD_TIME_TO_MSEC(syncErr));
}
if (std::abs(syncerror) > DVD_MSEC_TO_TIME(m_disconAdjustTimeMs))
// do corrections
```

Reply:
Code: https://github.com/fritsch/xbmc/commit/cd9e54f8a94e17b302de85c5e48bd71fb388def0

Reply:
Keep in mind: This was only tested on one device - the shield. There are more devices out there that might have certain side effects when video is starting. My key problem currently is: There might be setups where initially a 200 ms effect happens. But we don't correct it. We only correct it later on - after 30 seconds maybe, when the 100 is used as a fallback. Users with 200/300/400 ms off won't reach the 30 seconds ... as they feel something is really wrong.

Detecting such setups early and stating: Yes, the algorithm cannot find the issues is a good thing to not run into rproblems later.

Reply:
And: We have no hurry - the testing group can continue testing with the new method - there should not be any side effects, besides the algorithm can already finish with 100 after some few seconds - the same value it would reach only 30 seconds later anyways.

Reply:
As this is supposed to be a hardware/firmware/driver level latency, should the value be saved and used as the starting point?  Weighting over time would then dilute the potential impact from network issues, background threads, etc.

Reply:
> Keep in mind: This was only tested on one device - the shield. There are more devices out there that might have certain side effects when video is starting.

I agree with this, I haven't been able to test it on Windows yet due to lack of time. I don't expect any surprises here though: you'll probably always end up with a very similar 16 - 20 ms number that will work just fine anyway. If in any specific UHD Remux with a very high bitrate a higher number like 32 appears, it will also be fine.


> My key problem currently is: There might be setups where initially a 200 ms effect happens. But we don't correct it. We only correct it later on

I think your approach is wrong due these reasons:

**Proven fact 1:**
Sink returns completely unrealistic error values on Shield and Android in general. Therefore, these values cannot be trusted individually without some kind of previous filtering. Those errors do not reflect the a/v off-sync value of the stream. You can't fix the stream based on RAW errors because we'll be messing up a stream that is already OK without fixing.

**Proven fact 2:**
Most streams do NOT need to be a/v sync corrected because they are already correct. In other words, even removing all the a/v correction code, everything would still work fine in the vast majority of cases.

**Proven fact 3:**
In the worst case (true broken stream) that it needs to be fixed. This will happen the same as before because the C++ code after 30s is exactly the same as before. That is, all we do is hold up the correction because cannot be trust first initial error values returned by Sink.

You will agree with the 3 proven facts, right?

Now you want "correct" the stream based only in the first 100 ms value returned by sink!!!!

This is wrong => We should not trust the values returned by sink. Even less in the first ones received (hw not stabilized yet, high CPU spikes, HDD spinning up, file cache filling, USB 3.0 changing to performance mode, Wi-Fi switching to better channel due AI somewhere 😜, etc.)

If we "correct" the stream based on false/wrong data we will be spoiling the stream.

Conclusion: the change you propose kills the main idea of the PR and is no longer useful to resolve the initial problem.

**Option A:** Ignore a/v sync errors for 30s. We run the risk that during 30s an "rare true broken" stream is out of sync but it will be corrected later anyway (because C++ code is same after 30s).

**Option B**: We run the risk of "correcting" a stream that does not need to be corrected. We introduce artificial a/v off-sync errors that don't exist in the original stream. We introduce micro-stutters that are not necessary.

I prefer option **A**

**Conclusion 2:** In Android it is not possible to correct a/v sync errors quickly, it takes time to know which error threshold is due to "noise" and which is the true error of the stream.

Brilliant idea 1 ---> We can do a smoothed average of the error with 200 values. Yeah, right, great... but to have 200 values you need to wait a while anyway. With the first value of 100 ms you can't do anything except ignore it.

> Code: https://github.com/fritsch/xbmc/commit/cd9e54f8a94e17b302de85c5e48bd71fb388def0

In your commit I read: "Cancelling self-learning as error too high"
My mind thinks ==> "self-learning is very necessary as error too high (don't trust early error even if is too high)"

Reply:
> As this is supposed to be a hardware/firmware/driver level latency, should the value be saved and used as the starting point? Weighting over time would then dilute the potential impact from network issues, background threads, etc.

This number can be considered as a rating of how bad the hardware is but even if it returns 100 ms once it doesn't mean it's that bad all the time... it just means it can have bad 100 ms spikes.

Reply:
Okay - fine with me. It's master - so we can see how it works out and if we get complains.